src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -381,6 +381,41 @@
    "UPDATE factsets SET hash=md5(factsets.id::text)::bytea WHERE hash is NULL"
    "ALTER TABLE factsets ALTER COLUMN hash SET NOT NULL"))
 
+(defn add-support-for-historical-catalogs []
+  (jdbc/do-commands
+   "ALTER TABLE catalog_resources RENAME TO catalog_resources_tmp"
+   ;; CREATE certnames and catalog_resources transform tables
+   "CREATE TABLE catalog_resources (LIKE catalog_resources_tmp INCLUDING ALL)"
+   "CREATE TABLE latest_catalogs (catalog_id BIGINT REFERENCES catalogs(id) ON DELETE SET NULL, certname_id BIGINT PRIMARY KEY REFERENCES certnames(id) ON DELETE CASCADE)"
+   "ALTER TABLE catalog_resources DROP COLUMN catalog_id"
+   "ALTER TABLE catalog_resources ADD COLUMN certname_id BIGINT NOT NULL REFERENCES certnames(id) ON DELETE CASCADE"
+   "ALTER TABLE catalog_resources ADD PRIMARY KEY (certname_id, type, title)"
+
+   (str "INSERT INTO latest_catalogs"
+        "  (catalog_id, certname_id)"
+        "  SELECT catalogs.id, certnames.id"
+        "  FROM certnames"
+        "  LEFT JOIN catalogs ON catalogs.certname = certnames.certname")
+
+   (str "INSERT INTO catalog_resources"
+        "  (certname_id, resource, tags, type, title, exported, file, line)"
+        "  SELECT latest_catalogs.certname_id, cr.resource, cr.tags, cr.type, cr.title, cr.exported, cr.file, cr.line"
+        "  FROM catalog_resources_tmp cr"
+        "  LEFT JOIN latest_catalogs ON cr.catalog_id = latest_catalogs.catalog_id")
```

Comment:
I'm not following, why will a null `catalog_id` mean that there is no `certname_id`?